src/python/pants/backend/scala/bsp/rules.py

```diff
@@ -0,0 +1,120 @@
+    scala_version = scala.version_for_resolve(resolve)
+    return BuildTarget(
+        id=BuildTargetIdentifier(uri=f"pants:{target.address}"),
+        display_name=str(target.address),
+        base_directory=None,
+        tags=(),
+        capabilities=BuildTargetCapabilities(),
+        language_ids=("scala",),
+        dependencies=(),
+        data_kind="scala",
+        data=ScalaBuildTarget(
+            scala_organization="unknown",
+            scala_version=".".join(scala_version.split(".")[0:2]),
+            scala_binary_version=scala_version,
+            platform=ScalaPlatform.JVM,
+            jars=(),
+        ),
+    )
+
+
+@rule
+async def determine_scala_bsp_build_targets(
+    _: ScalaBSPBuildTargetsRequest,
+    all_scala_targets: AllScalaTargets,
+    jvm: JvmSubsystem,
+    scala: ScalaSubsystem,
+) -> BSPBuildTargets:
+    scala_bsp_build_targets = [
+        _pants_target_to_bsp_build_target(tgt, jvm, scala) for tgt in all_scala_targets
+    ]
+    return BSPBuildTargets(targets=tuple(scala_bsp_build_targets))
+
+
+# -----------------------------------------------------------------------------------------------
+# Scalac Options Request
+# See https://build-server-protocol.github.io/docs/extensions/scala.html#scalac-options-request
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class HandleScalacOptionsRequest:
+    bsp_target_id: BuildTargetIdentifier
+
+
+@dataclass(frozen=True)
+class HandleScalacOptionsResult:
+    item: ScalacOptionsItem
+
+
+@rule
+async def handle_bsp_scalac_options_request(
+    request: HandleScalacOptionsRequest,
+    build_root: BuildRoot,
+) -> HandleScalacOptionsResult:
+    uri = request.bsp_target_id.uri
+    if not uri.startswith("pants:"):
+        raise ValueError(f"Unknown URI for Pants BSP: {uri}")
+    raw_addr = uri[len("pants:") :]
+
+    # Verify the target exists by loading it. Exception will be thrown if it does not.
+    _ = await Get(WrappedTarget, AddressInput, AddressInput.parse(raw_addr))
+
+    return HandleScalacOptionsResult(
+        ScalacOptionsItem(
+            target=request.bsp_target_id,
+            options=(),
+            classpath=(),
+            # TODO: Figure out how to provide a classfiles output directory
+            class_directory=build_root.pathlib_path.joinpath("dist/bsp").as_uri(),
+        )
+    )
+
+
+@rule
+async def bsp_scalac_options_request(request: ScalacOptionsParams) -> ScalacOptionsResult:
+    results = await MultiGet(
+        Get(HandleScalacOptionsResult, HandleScalacOptionsRequest(btgt)) for btgt in request.targets
+    )
+    return ScalacOptionsResult(items=tuple(result.item for result in results))
+
+
+def rules():
+    return (
+        *collect_rules(),
+        UnionRule(BSPBuildTargetsRequest, ScalaBSPBuildTargetsRequest),
+        QueryRule(ScalacOptionsResult, (ScalacOptionsParams,)),
```

Comment:
Generally backends should not expose queries: either a Query is used by the engine itself in production (in which case initialization code should install it, á la https://github.com/pantsbuild/pants/blob/6309ae6224d4a0dd0e41d3feae6c49b58be74173/src/python/pants/init/engine_initializer.py#L272-L282), or it is used by a test (in which case a particular RuleRunner should). Otherwise, the rule graph calculation is doing computation it doesn't actually need (since no consumers will use that entrypoint).

Reply:
It was necessary to avoid rule graph errors at runtime. See the discussion at https://pantsbuild.slack.com/archives/C0D7TNJHL/p1644266141140719 where you suggested that I use `QueryRule` in this way. Is there a better place for these `QueryRule`'s?

Reply:
And can the bsp goal add these `QueryRule`'s somehow to the scheduler only when the `bsp` goal is invoked?

Reply:
> Is there a better place for these `QueryRule`'s?

For now, the place that I linked to in my comment above (in `src/python/pants/init/engine_initializer.py`). Formalizing that would be good, but it's something that only _non_-plugins are actually able to do (since a plugin cannot run queries: only install `@rules` that will satisfy queries).

> And can the `bsp` goal add these `QueryRule`'s somehow to the scheduler only when the `bsp` goal is invoked?

No. The rule graph is compiled once when the scheduler is created (generally: when `pantsd` starts up), so they need to be known ahead of time.

Reply:
Focusing on Scala-specific rules for a moment, putting the `QueryRule`'s for Scala-specific BSP methods into `engine_initalizer.py` would break the encapsulation offered by having backend plugins though. I.e., Pants core code ideally should not have any explicit knowledge of language-specific types.

I see that `UnionMembership` is initialized in `setup_graph_extended` immediately after the `QueryRule`'s are installed. It actually could be initialized before the `QueryRule`'s are installed and then queried for a (new) `BspHandlerMapping` type (which would have the input and output types). This would allow the engine initializer code to remain free of explicit knowledge of language backends.

Thoughts?

Reply:
> I see that `UnionMembership` is initialized in `setup_graph_extended` immediately after the `QueryRule`'s are installed. It actually could be initialized before the `QueryRule`'s are installed and then queried for a (new) `BspHandlerMapping` type (which would have the input and output types). This would allow the engine initializer code to remain free of explicit knowledge of language backends.

Yep, that sounds exactly right. But as mentioned elsewhere in here, also fine with punting.

Reply:
Will switch to using a union for the mapping definitions in a follow-on today. So punt for now.